OAuth: Add optional strict parsing of role_attribute_path

klausenbusk opened the pull request:
**What this PR does / why we need it**:

It is currently not possible to deny users access to Grafana when using [generic OAuth authentication](https://grafana.com/docs/grafana/latest/auth/generic-oauth/). 

**Which issue(s) this PR fixes**:

Fixes #23218

klausenbusk committed:
OAuth: Add strict role mapping

By default the user is assigned the role Viewer if role_attribute_path
doesn't return a role, which is not always desirable. This commit adds a
strict mode, which deny the user access if a role isn't returned.

Fix #26626

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/grafana?pullRequest=28021) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/grafana/grafana?pullRequest=28021) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/grafana?pullRequest=28021) it.</sub>

achatterjee-grafana reviewed (dismissed):
Thank you for contributing.

sakjur reviewed (dismissed):
(no changes requested, but I don't want this to be merged before we've discussed this a little)

I'd personally rather see a boolean that makes the parsing of roles strict and denies any non-role results 🤔
It seems bad that say someone accidentally writing `unauthorised` would give the users access.

So something like:

```
role_attribute_strict = true
```

meaning that anything except for `Viewer`, `Editor`, or `Admin` would be unauthorized.

alexanderzobnin commented:
@sakjur Yes, it definitely makes sense.

sakjur commented:
@klausenbusk What do you think about changing this to add an optional strict mode instead of hard-coding the deny access role?

klausenbusk commented:
> @klausenbusk What do you think about changing this to add an optional strict mode instead of hard-coding the deny access role?

It sounds like a better approach. I will try to implement it in the weekend (I don't have much time this week :/).

klausenbusk commented:
I haven't got time to test it (yet), but how does it looks?

sakjur commented:
Based on just looking at it it looks really good! Pretty much exactly like I'd do it myself.

Let me know when you've tested this and I'll test it on my end. After we've tested this, we'll also need to document the option and then I'd say it's pretty much ready for merge.

marefr commented:
@klausenbusk i closed your referenced issue as a duplicate of #23218. We've already closed the #22383 related to #23218.

@sakjur based on #22383 we need to discuss what to do with this one.

@klausenbusk we'll discuss this and come back with an answer.

sakjur commented:
@marefr :+1: I wasn't aware of https://github.com/grafana/grafana/pull/22383#issuecomment-606071605 I am a little concerned that we're making perfect to the enemy of the good here. I think this is a change for the better, making a feature (optionally) work the way it should've worked all along.

We're going to have to deprecate/support the `role_attribute_path` as we're moving to a better authentication solution. The difference between having a `role_attribute_strict` next to it as well isn't all too significant.

achatterjee-grafana reviewed (commented):
Looks good to me, but Dev will approve for merge.

marefr commented:
@sakjur lets pick this up after 7.3 stable has been released and discuss it

klausenbusk commented:
> @sakjur lets pick this up after 7.3 stable has been released and discuss it

Any update?

sakjur commented:
Yes, @marefr and I discussed this last week.

Our conclusion is that we (it's on my todo list) should update the documentation to make it clear that the default role is being used when an invalid role is returned and that that's something that might change in the future.

Auth in general is in a very messy state at the moment, and it is one of the areas we'd like to work on for 8.0, which means we want to take this PR into the context of that work. I'm still much in favor of adopting this behavior as I think it is the way it should work, but we want to make sure that regardless of if you use Grafana with OAuth, LDAP, vendor-specific auth, or whatnot we would like to handle it in the same way going forward. We don't want to further extend the gap between the different mechanisms at the moment (security rationale excluded — which is in a sense an edge-case for this particular feature, but still [in mine and Marcus's opinion] on the side of enhancement rather than bug fix).

In short, we won't be merging this PR for 7.4 to make sure we're not undoing it in 8.0 following soon after.

sakjur commented:
Being a bit further down the line with our internal discussions: this is a good idea and a good implementation thereof. Thank you for contributing to Grafana, @klausenbusk 🙏 

This PR looks good and can make it into Grafana 8.0 if its updated to be based on the most recent master and add documentation for the new option (for both Okta and Generic OAuth).

klausenbusk commented:
Rebased, added some documentation, squashed and I rewrote the commit message.

This is a reference in the Okta docs:
> Also, check Generic OAuth page for [JMESPath examples]({{< relref "generic-oauth.md/#jmespath-examples" >}}).

So I only described the strict mode in the generic oauth docs, perhaps it should be documented at both pages?

Not tested, but I think it is ready for review.

sakjur commented on docs/sources/auth/generic-oauth.md:
```diff
@@ -194,6 +194,8 @@ To ease configuration of a proper JMESPath expression, you can test/evaluate exp
 
 ### Role mapping
 
+By default the user is assigned the role `Viewer` if `role_attribute_path` doesn't return a role for the user, this is not always desirable and can be disabled by setting `role_attribute_strict = true`, which will deny the user access if no role is returned.
```
```suggestion
By default the user is assigned the role `Viewer` if `role_attribute_path` doesn't return a role for the user, this is not always desirable and can be disabled by setting `role_attribute_strict = true`, which will deny the user access if no role or an invalid role is returned.
```

I'm not reviewing the documentation, our docs squad will do that, but I think this might be worth mentioning 🙂

sakjur reviewed (approved):
Code looks good, I'll wait for someone in the documentations team to review the docs for this :+1:

klausenbusk committed:
Update docs/sources/auth/generic-oauth.md

Co-authored-by: Emil Tullstedt <emil.tullstedt@grafana.com>

achatterjee-grafana commented on docs/sources/auth/generic-oauth.md:
```diff
@@ -194,6 +194,8 @@ To ease configuration of a proper JMESPath expression, you can test/evaluate exp
 
 ### Role mapping
 
+By default the user is assigned the role `Viewer` if `role_attribute_path` doesn't return a role for the user, this is not always desirable and can be disabled by setting `role_attribute_strict = true`, which will deny the user access if no role or an invalid role is returned.
```
```suggestion
By default, if  the`role_attribute_path` does not return a role, the user is assigned the `Viewer` role. You can disable the role assignment by setting `role_attribute_strict = true`. It denies user access if no role or an invalid role is returned.
```

achatterjee-grafana reviewed (changes requested):
Added doc change request.

achatterjee-grafana committed:
Update docs/sources/auth/generic-oauth.md

Commit:
Update .gitignore file with WAN

Commit:
Removed WAN from .gitignore

Commit:
Merge branch 'master' into pr/28021

aocenas commented:
Hey @klausenbusk there is an issue with the build which is fixed on master. Can you pull the latest master, merge it with your branch and push that commit? That should fix the CI failure.

klausenbusk committed:
Merge remote-tracking branch 'upstream/master' into oauth-unauthorized-role

klausenbusk commented:
> Can you pull the latest master, merge it with your branch and push that commit?

Sure, done. I normally prefer rebasing but whatever works :)

burakksglu commented:
Glad to see this feature will come to Grafana in the future. We've used Keycloak to do this like mentioned in the #22383 but it wasn't the best solution. Thank you @klausenbusk & Grafana team!
